File: codes/go/chapter10/answer_A74.go
package main

import "fmt"

func main() {
	var n int
	fmt.Scan(&n)

	p := make([][]int, n+1)
	for i := 1; i <= n; i++ {
		p[i] = make([]int, n+1)
		for j := 1; j <= n; j++ {
			fmt.Scan(&p[i][j])
		}
	}

	x := make([]int, n+1)
	y := make([]int, n+1)

	for i := 1; i <= n; i++ {
		for j := 1; j <= n; j++ {
			if p[i][j] != 0 {
				x[i] = p[i][j]
				y[j] = p[i][j]
			}
		}
	}

	inversion_x := 0
	inversion_y := 0

	for i := 1; i <= n; i++ {
		for j := i + 1; j <= n; j++ {
			if x[i] > x[j] {
				inversion_x++
			}
			if y[i] > y[j] {
				inversion_y++
			}
		}
	}

	// 出力
	fmt.Println(inversion_x + inversion_y)
}

File: codes/ruby/1/chapter10/answer_A74.rb
n = gets.to_i
p = Array.new(n + 1) { Array.new(n + 1) }
(1..n).each do |i|
  p[i][1..n] = gets.split.map(&:to_i)
end

x = Array.new(n + 1)
y = Array.new(n + 1)
(1..n).each do |i|
  (1..n).each do |j|
    unless p[i][j].zero?
      x[i] = p[i][j]
      y[j] = p[i][j]
    end
  end
end

inversion_x = 0
inversion_y = 0
(1..n).each do |i|
  ((i + 1)..n).each do |j|
    inversion_x += 1 if x[i] > x[j]
    inversion_y += 1 if y[i] > y[j]
  end
end

# 出力
puts inversion_x + inversion_y
